File: 1441. Build an Array With Stack Operations/solution.py

from typing import List


# 1 2 3 
# [1,3]
# [1, 3]
class Solution:
    def buildArray(self, target: List[int], n: int) -> List[str]:
        ans = []
        index = 0
        shift = 0

        for i in range(1,n+1): 
            if i - 1 == target[-1]: 
                return ans

            if i == target[index]: 
                while shift: 
                    ans.append("Pop")
                    shift -= 1
                index += 1
            else: 
                shift += 1

            ans.append("Push")
        return ans

File: 1441. Build an Array With Stack Operations/solution.cpp
#include <string>
#include <vector>


using namespace std;

class Solution {
public:
    vector<string> buildArray(vector<int>& target, int n) {

        int shift = 0, index = 0;
        vector<string> ans;
        for (int i = 1; i <= n; i++) {
            if (i - 1 == target[target.size() - 1]) {
                return ans;
            }

            if (target[index] == i) {
                while (shift) {
                    ans.push_back("Pop");
                    shift--;
                }

                index++;
            } else
                shift++;

            ans.push_back("Push");
        }

        return ans;
    }
};
